### PR TITLE
initial trace module

### PR DESCRIPTION
**Pull request by tsloughter**
Thoughts?

### CODE REVIEW

File: lib/opencensus/trace.ex [hunk 1]
```diff
@@ -0,0 +1,27 @@
+defmodule Opencensus.Trace do
+  @doc "Wraps the given block in a tracing child span with the given label/name"
+  defmacro with_child_span(label, do: block) do
+    quote do
+      :ocp.with_child_span(unquote(label), %{})
+
+      try do
+        unquote(block)
+      after
+        :ocp.finish_span()
```
Comment (hauleth) [inline lib/opencensus/trace.ex]: Use `:oc_trace.finish_span/1` to be sure that the span you opened on L5 is the same that you close there.

File: lib/opencensus/trace.ex [hunk 2]
```diff
@@ -0,0 +1,27 @@
+defmodule Opencensus.Trace do
+  @doc "Wraps the given block in a tracing child span with the given label/name"
+  defmacro with_child_span(label, do: block) do
```
Comment (hauleth) [inline lib/opencensus/trace.ex]: I would probably use function instead of macro.

File: lib/opencensus/trace.ex [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      try do
+        unquote(block)
+      after
+        :ocp.finish_span()
+      end
+    end
+  end
+
+  @doc "Wraps the given block in a tracing child span with the given label/name and additional attributes"
+  defmacro with_child_span(label, attributes, do: block) do
```
Comment (hauleth) [inline lib/opencensus/trace.ex]: Both cases can be squashed into single definition by using:

```elixir
defmacro with_child_span(label, attributes \\ quote do: %{}, do: block) do
  # …
```

**Review by hauleth — CHANGES_REQUESTED** [review]
I would be wary of using macros and unless you wan to do anything in compile time then I would avoid them.

File: lib/opencensus/trace.ex (hunk 1 above)
Comment (tsloughter) [reply lib/opencensus/trace.ex]: Ah yea, I'll fix this up.

**Comment by tsloughter** [comment]
I like the macro, and usually I hate macros.

**Comment by hauleth** [comment]
But there is no point in using macro except making it slightly "nicer" by using `do … end` syntax. For me it is pointless and potentially can cause weird error reports. It would make sense if you would provide context as the attributes. That is why I have created census-instrumentation/opencensus-erlang#112, in that case it would make sense ;)

File: lib/opencensus/trace.ex (hunk 3 above)
Comment (tsloughter) [reply lib/opencensus/trace.ex]: That doesn't compile so I'm keeping it as two separate definitions for now.

Comment (hauleth) [reply lib/opencensus/trace.ex]: @tsloughter `quote(do: [])`, alternatively you can do:

```elixir
defmacro with_child_span(label, do: block),
  do: with_child_span(label, quote(do: []), do: block)
```

**Review by hauleth — APPROVED** [review]

**Pull request merged**